service/src/main/java/org/ehrbase/plugin/AbstractPluginAspect.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2022. vitasystems GmbH and Hannover Medical School.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.ehrbase.plugin;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.function.TriFunction;
+import org.aspectj.lang.ProceedingJoinPoint;
+import org.aspectj.lang.annotation.Pointcut;
+import org.ehrbase.api.exception.InternalServerException;
+import org.ehrbase.plugin.extensionpoints.CompositionExtensionPoint;
+import org.springframework.beans.factory.ListableBeanFactory;
+import org.springframework.core.annotation.AnnotationAwareOrderComparator;
+
+/**
+ * @author Stefan Spiska
+ */
+/*TODO: maybe we should add the Ordered interface, so derived aspects will be evaluated in a predictable order in
+   relation to other aspects we might have or add later for other features or even through a plugin?*/
+public abstract class AbstractPluginAspect<EXTENSIONPOINT> {
+
+  private final Comparator<Map.Entry<String, EXTENSIONPOINT>> EXTENSION_POINTS_COMPARATOR =
+      // respect @Order
+      ((Comparator<Map.Entry<String, EXTENSIONPOINT>>)
+          (e1, e2) ->
+              AnnotationAwareOrderComparator.INSTANCE.compare(e1.getValue(), e2.getValue()))
+          .reversed()
+          // ensure constant ordering
+          .thenComparing(Map.Entry::getKey);
+
+  protected final ListableBeanFactory beanFactory;
+
+  private final Class<EXTENSIONPOINT> clazz;
+
+  protected AbstractPluginAspect(ListableBeanFactory beanFactory, Class<EXTENSIONPOINT> clazz) {
+    this.beanFactory = beanFactory;
+    this.clazz = clazz;
+  }
+
+  @Pointcut("within(org.ehrbase.service..*)")
+  public void inServiceLayerPC() {}
+
+  /**
+   * Proceed with Error handling.
+   *
+   * @param pjp
+   * @param args
+   * @return
+   */
+  protected Object proceed(ProceedingJoinPoint pjp, Object[] args) {
+    try {
+      return pjp.proceed(args);
+    } catch (RuntimeException e) {
+      // Simple rethrow to handle in Controller layer
+      throw e;
+    } catch (Throwable e) {
+      // should never happen
+      throw new InternalServerException(e.getMessage(), e);
```

Comment:
I think it's no good idea to potentially degrade an Error to a RuntimeException.
Should we rethrow  Errors?

Reply:
i agree. intellij seems to have optimized that away because i had the wrong order of the catch blocks or something